Assert called

SukiCZ opened the pull request:
Hello :wave: 

This adds `assert_called` and other unittest-like methods to `aioresponses`.

Resolves: #132

What are your thoughts?

Cheers :beers:

pnuckowski commented on aioresponses/core.py:
```diff
@@ -298,19 +348,86 @@ def add(self, url: 'Union[URL, str, Pattern]', method: str = hdrs.METH_GET,
+        """
+        if len(self._calls) != 0:
+            msg = ("Expected '%s' to not have been called. Called %s times."
+                   % (self.__class__,
```
`self.__class__.__name__`

pnuckowski commented on aioresponses/core.py:
```diff
@@ -234,6 +278,8 @@ def wrapped(*args, **kwargs):
     def clear(self):
         self._responses.clear()
         self._matches.clear()
+        self._call_args = None
```
the name of that variable suggest that this is call arguments, but type annotation is `RequestMatch` :)

pnuckowski commented on aioresponses/core.py:
```diff
@@ -298,19 +348,86 @@ def add(self, url: 'Union[URL, str, Pattern]', method: str = hdrs.METH_GET,
+        self.assert_called()
+
+        expected = self._call_matcher((args, kwargs))
+        actual = self._call_matcher(self._call_args)
```
Scenario. 
User mocks two or more requests (`http://example.com` and `http://httpbin.org/`)
What I would expected is to successfully call 
`assert_called_with('http://example.com')` 
and  
`assert_called_with('http://httpbin.org/')`

With current implementation this is not possible :/

SukiCZ replied:
I've added list of calls `_request_match_list` and method `assert_any_call` that checks if any call like that was made. You can find a test for this functionality in `test_assert_any_call_success` :ok_hand:

pnuckowski commented on aioresponses/core.py:
```diff
@@ -201,6 +243,8 @@ def __init__(self, **kwargs):
                              side_effect=self._request_mock,
                              autospec=True)
         self.requests = {}
+        self._calls = CallList()
```
We already have `requests` collection where we "keep" some call breadcrumbs. 
Maybe we should use/extend that in order to implement those new methods.

SukiCZ replied:
Hey @pnuckowski :wave: :smiley: 
Current implementation has:
* `requests` as a dictionary of `{'key': [RequestCall(...),]}`
* `_responses` as a dictionary of `{'key': ClientResponse(...)}`
* `_matches` as a list of `RequestMatch(...)` but the items are deleted after match

My idea here was to keep `RequestMatch`es somewhere for later evaluation.

I'm afraid to change behavior of existing properties and `calls` makes it more straightforward to me. My only idea is to make them more private... What would you recommend here?

Thanks!!

pnuckowski commented on aioresponses/core.py:
```diff
@@ -247,6 +293,10 @@ def stop(self) -> None:
         self.clear()
 
+    @property
+    def calls(self):
```
This comment relates [to this comment](https://github.com/pnuckowski/aioresponses/pull/133/files#r330274980)

I do not want to "expose" new property, because it will be tempting (for users) to use them. 
In future the way how aioresponses performs `assert_called_once()` and other `assert_*` methods might change (different structures, etc) and I do now want to "break" compatibility :)

pnuckowski commented:
Hi @SukiCZ 
I have added some comments to your PR. Good job, might be merged but I believe that we can do more with this :) 
Let me know if You have time to work on this, if not - let me know either (in that case I will merge that PR and open new one with suggestions)

PS. sorry for late review. Lots of stuff was going on (marriage etc)

cancan101 commented:
Any updates on this getting merged?

UncleGoogle commented:
@SukiCZ that would be pity to have your work lost. We need this feature, could you please take a look?

SukiCZ commented:
Hey all, thanks for reminding me this! 🎉 

I'd review my changes and implemented @pnuckowski 's idea that another collection is not required.
Updated changes are depending of `requests` collection and poorly match requests only by `method` and `url` - which left a `#TODO` mark in `assert_called_with` for specific parameters.

I believe better collection would be more suitable but basic functionality still works (by `method` and `url`) / Kept the unit-tests 🔝

SukiCZ commented:
* Fixed the `TODO` 🐛
* extended unit-test for `POST` with `data`/`payload`/`headers` [link](https://github.com/pnuckowski/aioresponses/pull/133/commits/a8eddb865bbdb0816d974a0ec63bdb62bfcf7f91#diff-a2fed4bfc58d957c79fb1eb4d67c9dd6142068ff3020b5be4a93a28cf0fcb549R184) 🧪
* added annotation for static typing 📝
* `tox` and `flake8` passing ✅

Feature should be ready from my side
Thank you @pnuckowski, I could work with `async` again
[Sorry I took so long](https://www.youtube.com/watch?v=gdk8g22A9bM)

UncleGoogle commented:
@pnuckowski could you take a look? 
(soft ping in case you do not receive notifications otherwise)

SukiCZ commented on README.rst:
```diff
@@ -112,6 +114,7 @@ for convenience use *payload* argument to mock out json response. Example below.
         # note that we pass 'connection' but get 'Connection' (capitalized)
         # under the neath `multidict` is used to work with HTTP headers
         assert resp.headers['Connection'] == 'keep-alive'
+        m.assert_called_once_with('http://example.com', method='POST')
```
I've missed a `headers` here

gaby commented:
@pnuckowski Cam you take a look at this, its a must need feature.

Thanks!

pnuckowski commented on aioresponses/core.py:
```diff
@@ -331,6 +331,113 @@ def add(self, url: 'Union[URL, str, Pattern]', method: str = hdrs.METH_GET,
+        """assert that the mock was called once with the specified arguments.
+        Raises an AssertionError if the args and keyword args passed in are
+        different to the only call to the mock."""
+        self.assert_called_once()
```
I have doubts about this. 
assert_called_once_with - should assert if it was called once for specific criteria(url method)

But I don"t want to force and let the ppl/users help to "define" what form it should take. 

But for now lets ship it :)
